internal/cri/server/events.go

```diff
@@ -186,6 +190,33 @@ func (c *criService) handleContainerExit(ctx context.Context, e *eventtypes.Task
 			return fmt.Errorf("failed to load task for container: %w", err)
 		}
 	} else {
+		if e.ExitStatus == oomExitCodeInLinux && cntr.Status.Get().Reason != oomExitReason {
+			var oomKilled bool
+
+			// NOTE(fuweid):
+			//
+			// 1. Even if the container’s main process is not OOM-killed, an OOM kill of an exec
+			//    process or a child process can still update status.Reason to OOMKilled.
+			//    This appears to be a CRI design issue and is worth highlighting.
+			//
+			// 2. The CRI plugin handles all OOM events in a single goroutine. This can lead to
+			//    a race condition where the Wait response is returned before the OOM event is
+			//    processed. As a result, the client may first observe status.Reason as Error,
+			//    and only later see it updated to OOMKilled. This behavior is inconsistent.
+			//    To avoid this, we should check metrics to determine whether an OOM event
+			//    occurred before reporting the final status.
+			oomKilled, err = c.oomEventOccurred(ctx, sandboxID, task)
+			if err == nil && oomKilled {
+				err = cntr.Status.UpdateSync(func(status containerstore.Status) (containerstore.Status, error) {
+					status.Reason = oomExitReason
```

Comment:
kk.. then down in the following oom event code we can skip the update sync when we've already set the reason in the container task exit 
```
	case *eventtypes.TaskOOM:
		log.L.Infof("TaskOOM event %+v", e)
		// For TaskOOM, we only care which container it belongs to.
		cntr, err := ce.c.containerStore.Get(e.ContainerID)
		if err != nil {
			if !errdefs.IsNotFound(err) {
				return fmt.Errorf("can't find container for TaskOOM event: %w", err)
			}
			return nil
		}
		err = cntr.Status.UpdateSync(func(status containerstore.Status) (containerstore.Status, error) {
			status.Reason = oomExitReason
			return status, nil
		})
```

Reply:
Yeah. Updated. No need to make update on that with extra IO